crates/aquascope/src/analysis/permissions/flow.rs

```diff
@@ -0,0 +1,474 @@
+  fx::{FxHashMap as HashMap, FxHashSet as HashSet},
+  graph::{
+    scc::Sccs, vec_graph::VecGraph, DirectedGraph, WithNumNodes, WithSuccessors,
+  },
+};
+use rustc_index::{
+  bit_set::{BitMatrix, HybridBitSet},
+  vec::Idx,
+};
+use serde::Serialize;
+use ts_rs::TS;
+
+use super::{places_conflict, Origin, PermissionsCtxt};
+use crate::mir::utils::{BodyExt, ToRegionVid};
+
+rustc_index::newtype_index! {
+  pub struct SccIdx {
+    DEBUG_FORMAT = "scc{}"
+  }
+}
+
+impl polonius_engine::Atom for SccIdx {
+  fn index(self) -> usize {
+    rustc_index::vec::Idx::index(self)
+  }
+}
+
+#[derive(Debug, Clone, Serialize, TS)]
+#[ts(export)]
+pub enum FlowEdgeKind {
+  /// A local value is flowing into an abstract region.
+  LocalOutlivesUniversal,
+
+  /// An abstract region is flowing into another,
+  /// but there is no constraint between the two.
+  MissingUniversalConstraint,
+
+  /// A local value will be invalidated at an exit point.
+  ///
+  /// Note, this case is not disjoint from `LocalOutlivesUniversal` but
+  /// is slightly more generic. This could be returned if a local value escapes
+  /// its scope, not necessarily a function boundary, which would be necessary
+  /// for `LocalOutlivesUniversal`. Because it's more generic, this type of
+  /// edge kind would be reported with lower priorty.
+  LocalInvalidatedAtExit,
+  Ok,
+}
+
+impl FlowEdgeKind {
+  pub fn is_valid_flow(&self) -> bool {
+    matches!(self, FlowEdgeKind::Ok)
+  }
+}
+#[allow(dead_code)]
+pub struct RegionFlows {
+  /// The flow constraint graph over the `subset_base` relation.
+  constraint_graph: Sccs<Origin, SccIdx>,
+
+  /// Full set of known flows per the `known_placeholder_subset` relation.
+  known_flows: BitMatrix<SccIdx, SccIdx>,
+
+  /// Local regions that could dangle due to an exit invalidation.
+  dangling_local_sources: HybridBitSet<SccIdx>,
+
+  /// The set of abstract components that a given component could contain.
+  contains_abstract: BitMatrix<SccIdx, SccIdx>,
+
+  /// The set of local components that a given component could contain.
+  contains_local: BitMatrix<SccIdx, SccIdx>,
+}
+
+impl RegionFlows {
+  pub fn scc(&self, origin: Origin) -> SccIdx {
+    self.constraint_graph.scc(origin)
+  }
+
+  pub fn has_abstract_member(&self, origin: Origin) -> bool {
+    self.contains_abstract.count(self.scc(origin)) > 0
+  }
+
+  pub fn has_local_member(&self, origin: Origin) -> bool {
+    self.contains_local.count(self.scc(origin)) > 0
+  }
+
+  /// Get the specific kind of flow edge that connects `from` and `to`.
+  #[allow(clippy::match_same_arms)]
+  pub(crate) fn flow_kind(&self, from: Origin, to: Origin) -> FlowEdgeKind {
+    let scc_from = self.constraint_graph.scc(from);
+    let scc_to = self.constraint_graph.scc(to);
+
+    // Data can always flow within the same SCC.
+    if scc_from == scc_to {
+      return FlowEdgeKind::Ok;
+    }
+
+    let from_contains_abstract = self.contains_abstract.count(scc_from) > 0;
```

Comment:
I would replace these with the corresponding `self.has_(abstract|local)_member` call. The performance hit of calling `self.scc` should be negligible.